### PR TITLE
Speed up ntuple (fixes #17703)

### LINKED ISSUE #17703: surprising memory allocation in ntuple
`ntuple` seems to allocate a lot of memory compared to a naive definition:
```jl
_ntuple(f::Function, n) = ((f(i) for i in 1:n)...)

@time ntuple(identity, 1000);
@time _ntuple(identity, 1000);
```
gives
```
  0.003675 seconds (24.75 k allocations: 2.313 MB)
  0.000101 seconds (504 allocations: 39.891 KB)
```

### PR DESCRIPTION
**Pull request by p-zubieta**

### CODE REVIEW

**Comment by timholy** [comment]
LGTM

**Comment by pabloferz** [comment]
As a note: I wanted to do the same for map instead of the current implementation that uses an `Array{Any}`, but I think that the fact that there is no SIMDilability for generators (yet) won't let a strategy like this one to beat the one already in Base.

**Comment by JeffBezanson** [comment]
Unfortunately there's going to be an `Array{Any}` as part of the implementation of `...`. It has to collect all the arguments somewhere to do dispatch.

**Comment by timholy** [comment]
Actually, it looks like `ntuple` is under-tested. Would you mind adding a test for each branch? e.g.,
```jl
for n = -1:20
    t = ntuple(identity, n)
    for i = 1:n
        @test t[i] == i
    end
end
```

File: base/tuple.jl [hunk 1]
```diff
@@ -63,13 +63,16 @@ end
...
 
 ntuple(f::Function, n::Integer) =
-    n<=0 ? () :
-    n==1 ? (f(1),) :
-    n==2 ? (f(1),f(2),) :
-    n==3 ? (f(1),f(2),f(3),) :
-    n==4 ? (f(1),f(2),f(3),f(4),) :
-    n==5 ? (f(1),f(2),f(3),f(4),f(5),) :
-    tuple(ntuple(f,n-5)..., f(n-4), f(n-3), f(n-2), f(n-1), f(n))
+    n <= 0 ? () :
```
Comment (KristofferC) [inline base/tuple.jl]: Not related to the PR but shouldnt there be an error for n < 0`?

**Pull request merged**